src/tools/profile/read_profile.c

```diff
@@ -131,49 +167,76 @@ static double time_to_usec(profile_data_t *data, options_t *opts, uint64_t time)
 {
-    const ucs_profile_location_t *loc1 = l1;
-    const ucs_profile_location_t *loc2 = l2;
+    const ucs_profile_thread_location_t *loc1 = l1;
+    const ucs_profile_thread_location_t *loc2 = l2;
     return (loc1->total_time > loc2->total_time) ? -1 :
            (loc1->total_time < loc2->total_time) ? +1 :
            0;
 }
 
-static void show_profile_data_accum(profile_data_t *data, options_t *opts)
+static int show_profile_data_accum(profile_data_t *data, options_t *opts)
 {
-    uint32_t num_locations = data->header->num_locations;
-    ucs_profile_location_t *sorted_locations;
-    ucs_profile_location_t *loc;
+    const uint32_t            num_locations          = data->header->num_locations;
+    profile_sorted_location_t *sorted_locations      = NULL;
+    const ucs_profile_thread_location_t *thread_location;
+    profile_sorted_location_t *sorted_loc;
+    const ucs_profile_location_t *loc;
+    unsigned location_idx;
+    int ret;
 
-    sorted_locations = malloc(sizeof(*sorted_locations) * num_locations);
+    sorted_locations = calloc(sizeof(*sorted_locations), num_locations);
     if (sorted_locations == NULL) {
-        return;
+        ret = -ENOMEM;
+        goto out;
+    }
+
+    /* Go over the list of threads provided by the user and accumulate the times
+     * and counts from all threads. In addition, track which calls were made from
+     * which threads.
+     */
+    for (location_idx = 0; location_idx < num_locations; ++location_idx) {
+        sorted_loc               = &sorted_locations[location_idx];
+        thread_location          = &data->thread.locations[location_idx];
+        sorted_loc->location_idx = location_idx;
+        sorted_loc->count       += thread_location->count;
+        sorted_loc->total_time  += thread_location->total_time;
     }
 
     /* Sort locations */
-    memcpy(sorted_locations, data->locations, sizeof(*sorted_locations) * num_locations);
-    qsort(sorted_locations, num_locations, sizeof(*sorted_locations), compare_locations);
+    qsort(sorted_locations, num_locations, sizeof(*sorted_locations),
+          compare_locations);
 
     /* Print locations */
     printf("%*s %13s %13s %10s                FILE     FUNCTION\n",
            FUNC_NAME_MAX_LEN, "NAME", "AVG", "TOTAL", "COUNT");
-    for (loc = sorted_locations; loc < sorted_locations + num_locations; ++loc) {
+
+    for (sorted_loc = sorted_locations;
+         sorted_loc < sorted_locations + num_locations; ++sorted_loc) {
+
+        if (sorted_loc->count == 0) {
+            continue;
+        }
+
+        loc = &data->locations[sorted_loc->location_idx];
+
         switch (loc->type) {
         case UCS_PROFILE_TYPE_SAMPLE:
             printf("%*.*s %13s %13s %10ld %18s:%-4d %s()\n",
                    FUNC_NAME_MAX_LEN, FUNC_NAME_MAX_LEN,
                    loc->name,
                    "-",
                    "-",
-                   (long)loc->count,
+                   (long)sorted_loc->count,
```

Comment:
what about to use `%10zu` instead of `%10ld`?
or make `long count` instead of `size_t long`?